File: src/sort/DutchFlagProblem.java
package sort;

import java.util.Arrays;

public class DutchFlagProblem {
    /**
     * 给定区间内的数, 按照给定轴, 排成[&lt;, =, &gt;]
     *
     * @param arr 目标数组
     * @param l   左边界
     * @param r   右边界, 同时也是轴(目标数)
     * @return 返回等于区域的左右边界
     */
    public static int[] dutchFlag(int[] arr, int l, int r) {
        if (l > r) return new int[]{-1, -1};
        if (l == r) return new int[]{l, r};
        int lt = l - 1, gt = r, index = l;
        while (index < gt) {
            if (arr[index] < arr[r]) {
                swap(arr, ++lt, index++);
            } else if (arr[index] > arr[r]) {
                swap(arr, --gt, index);
            } else {
                index++;
            }
        }
        swap(arr, gt, r);
        return new int[]{lt + 1, gt};
    }

    public static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    public static void main(String[] args) {
        int[] arr = new int[]{8, 7, 32, 2, 2, 1, 3, 12, 11, 12, 7};
        System.out.println(Arrays.toString(dutchFlag(arr, 0, arr.length - 1)));
        System.out.println(Arrays.toString(arr));
    }
}

File: src/sort/dutch_flag_problem.py
"""荷兰国旗问题
将给定区间内的数, 按照给定轴, 排成[<; =; >;]三个区域
"""


def dutch_flag(arr: list[int], l: int, r: int) -> tuple[int, int]:
    """将[l, r]范围上的数, 以r索引位置为轴, 排成[<r; =r; >r]
    :return: 返回等于区域的左右边界, 包含等于区域
    """
    if l == r:
        return l, r
    lt, gt, index = l - 1, r, l
    while index < gt:
        # 当前值小于轴, 将小于区的下一个与当前值交换, 当前继续
        if arr[index] < arr[r]:
            lt += 1
            arr[lt], arr[index] = arr[index], arr[lt]
            index += 1
        # 当前值大于轴, 将大于区域的前一项与当期值交换, 同时大于区向前移
        elif arr[index] > arr[r]:
            gt -= 1
            arr[index], arr[gt] = arr[gt], arr[index]
        else:
            index += 1
    arr[gt], arr[r] = arr[r], arr[gt]
    return lt + 1, gt


if __name__ == '__main__':
    arr = [8, 7, 32, 2, 2, 1, 3, 12, 11, 12, 7]
    print(dutch_flag(arr, 0, len(arr) - 1))
    print(arr)
